File: src/feishu_auth_kit/cardkit.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from .agent_runtime import AgentEvent, AgentTurnResult
from .message_context import FeishuMessageContext


@dataclass(frozen=True)
class CardKitStep:
    id: str
    kind: str
    title: str
    status: str
    detail: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "id": self.id,
            "kind": self.kind,
            "title": self.title,
            "status": self.status,
            "detail": self.detail,
            "metadata": self.metadata,
        }
        return {key: value for key, value in payload.items() if value not in (None, {})}


@dataclass(frozen=True)
class SingleCardRun:
    runner: str
    message_id: str | None
    chat_id: str | None
    sender_open_id: str | None
    status: str
    summary: str
    final_text: str
    steps: list[CardKitStep]
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema": "feishu-auth-kit.cardkit.single_card.v1",
            "type": "agent_run_single_card",
            "runner": self.runner,
            "message_id": self.message_id,
            "chat_id": self.chat_id,
            "sender_open_id": self.sender_open_id,
            "status": self.status,
            "summary": self.summary,
            "final_text": self.final_text,
            "steps": [step.to_dict() for step in self.steps],
            "metadata": self.metadata,
        }


def _event_detail(event: AgentEvent) -> str | None:
    if event.detail:
        return event.detail
    if event.tool_input is not None:
        if isinstance(event.tool_input, str):
            return event.tool_input
        return json.dumps(event.tool_input, ensure_ascii=False, sort_keys=True)
    return event.text


def _step_title(event: AgentEvent) -> str:
    if event.kind == "tool_call":
        return f"Tool: {event.tool_name or 'unknown'}"
    if event.kind == "tool_result":
        return f"Tool result: {event.tool_name or 'unknown'}"
    if event.kind == "start":
        return "Runner started"
    if event.kind == "running":
        return "Runner running"
    if event.kind == "completed":
        return "Runner completed"
    if event.kind == "error":
        return "Runner error"
    if event.kind == "stderr_warning":
        return "Runner warning"
    if event.kind == "assistant_message":
        return "Assistant response"
    return event.text or "Runtime status"


def _step_from_event(index: int, event: AgentEvent) -> CardKitStep:
    return CardKitStep(
        id=f"step-{index}",
        kind=event.kind,
        title=_step_title(event),
        status=event.state or "completed",
        detail=_event_detail(event),
        metadata=event.metadata,
    )


def _summary(text: str, *, max_chars: int = 160) -> str:
    compact = " ".join(text.strip().split())
    if len(compact) <= max_chars:
        return compact
    return compact[: max_chars - 1].rstrip() + "…"


def build_single_card_run(
    context: FeishuMessageContext,
    result: AgentTurnResult,
) -> SingleCardRun:
    steps = [_step_from_event(index, event) for index, event in enumerate(result.events, start=1)]
    return SingleCardRun(
        runner=result.runner,
        message_id=context.message_id,
        chat_id=context.chat_id,
        sender_open_id=context.sender_open_id,
        status=result.status,
        summary=_summary(result.output_text),
        final_text=result.output_text,
        steps=steps,
        metadata={
            "event_id": context.event_id,
            "event_type": context.event_type,
            **result.metadata,
        },
    )

File: src/cardkit.ts
import { AgentEvent, AgentTurnResult } from './agent-runtime.js';
import { FeishuMessageContext } from './message-context.js';

export interface CardKitStepOptions {
  id: string;
  kind: string;
  title: string;
  status: string;
  detail?: string | null;
  metadata?: Record<string, any>;
}

export class CardKitStep {
  readonly id: string;
  readonly kind: string;
  readonly title: string;
  readonly status: string;
  readonly detail: string | null;
  readonly metadata: Record<string, any>;

  constructor(options: CardKitStepOptions) {
    this.id = options.id;
    this.kind = options.kind;
    this.title = options.title;
    this.status = options.status;
    this.detail = options.detail ?? null;
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }

  toDict(): Record<string, any> {
    const payload: Record<string, any> = {
      id: this.id,
      kind: this.kind,
      title: this.title,
      status: this.status,
      detail: this.detail,
      metadata: this.metadata,
    };
    const result: Record<string, any> = {};
    for (const [key, value] of Object.entries(payload)) {
      if (value === null || value === undefined) {
        continue;
      }
      if (
        typeof value === 'object' &&
        !Array.isArray(value) &&
        Object.keys(value).length === 0
      ) {
        continue;
      }
      result[key] = value;
    }
    return result;
  }
}

export interface SingleCardRunOptions {
  runner: string;
  messageId: string | null;
  chatId: string | null;
  senderOpenId: string | null;
  status: string;
  summary: string;
  finalText: string;
  steps: CardKitStep[];
  metadata?: Record<string, any>;
}

export class SingleCardRun {
  readonly runner: string;
  readonly messageId: string | null;
  readonly chatId: string | null;
  readonly senderOpenId: string | null;
  readonly status: string;
  readonly summary: string;
  readonly finalText: string;
  readonly steps: CardKitStep[];
  readonly metadata: Record<string, any>;

  constructor(options: SingleCardRunOptions) {
    this.runner = options.runner;
    this.messageId = options.messageId;
    this.chatId = options.chatId;
    this.senderOpenId = options.senderOpenId;
    this.status = options.status;
    this.summary = options.summary;
    this.finalText = options.finalText;
    this.steps = [...options.steps];
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }

  toDict(): Record<string, any> {
    return {
      schema: 'feishu-auth-kit.cardkit.single_card.v1',
      type: 'agent_run_single_card',
      runner: this.runner,
      message_id: this.messageId,
      chat_id: this.chatId,
      sender_open_id: this.senderOpenId,
      status: this.status,
      summary: this.summary,
      final_text: this.finalText,
      steps: this.steps.map((step) => step.toDict()),
      metadata: this.metadata,
    };
  }
}

function eventDetail(event: AgentEvent): string | null {
  if (event.detail) {
    return event.detail;
  }
  if (event.toolInput !== null && event.toolInput !== undefined) {
    if (typeof event.toolInput === 'string') {
      return event.toolInput;
    }
    return JSON.stringify(event.toolInput);
  }
  return event.text;
}

function stepTitle(event: AgentEvent): string {
  if (event.kind === 'tool_call') {
    return `Tool: ${event.toolName || 'unknown'}`;
  }
  if (event.kind === 'tool_result') {
    return `Tool result: ${event.toolName || 'unknown'}`;
  }
  if (event.kind === 'start') {
    return 'Runner started';
  }
  if (event.kind === 'running') {
    return 'Runner running';
  }
  if (event.kind === 'completed') {
    return 'Runner completed';
  }
  if (event.kind === 'error') {
    return 'Runner error';
  }
  if (event.kind === 'stderr_warning') {
    return 'Runner warning';
  }
  if (event.kind === 'assistant_message') {
    return 'Assistant response';
  }
  return event.text || 'Runtime status';
}

function stepFromEvent(index: number, event: AgentEvent): CardKitStep {
  return new CardKitStep({
    id: `step-${index}`,
    kind: event.kind,
    title: stepTitle(event),
    status: event.state || 'completed',
    detail: eventDetail(event),
    metadata: event.metadata,
  });
}

function makeSummary(text: string, maxChars = 160): string {
  const compact = text.trim().split(/\s+/).join(' ');
  if (compact.length <= maxChars) {
    return compact;
  }
  return compact.slice(0, maxChars - 1).trimEnd() + '…';
}

export function buildSingleCardRun(
  context: FeishuMessageContext,
  result: AgentTurnResult,
): SingleCardRun {
  const steps = result.events.map((event, i) => stepFromEvent(i + 1, event));
  return new SingleCardRun({
    runner: result.runner,
    messageId: context.messageId ?? null,
    chatId: context.chatId ?? null,
    senderOpenId: context.senderOpenId ?? null,
    status: result.status,
    summary: makeSummary(result.outputText),
    finalText: result.outputText,
    steps,
    metadata: {
      event_id: context.eventId,
      event_type: context.eventType,
      ...result.metadata,
    },
  });
}
